### PR TITLE
Everywhere: Push aarch64/ARM64 little bit further

### PR DESCRIPTION
**Pull request by konradekk**

### CODE REVIEW

File: Userland/Libraries/CMakeLists.txt [hunk 1]
```diff
@@ -63,4 +63,5 @@ add_subdirectory(LibWeb)
 add_subdirectory(LibWebSocket)
 add_subdirectory(LibWebView)
 add_subdirectory(LibX86)
+add_subdirectory(LibARM64)
```
Comment (Beyley) [inline Userland/Libraries/CMakeLists.txt:66]: This should be alphabetically sorted

File: Kernel/Process.cpp [hunk 2]
```diff
@@ -246,8 +246,8 @@ ErrorOr<NonnullLockRefPtr<Process>> Process::try_create_user_process(LockRefPtr<
     }));
 
     Thread* new_main_thread = nullptr;
-    u32 prev_flags = 0;
```
Comment (FireFox317) [inline Kernel/Process.cpp:249]: I've called this concept `previous_interrupts_state` before, see https://github.com/SerenityOS/serenity/blob/00db1dffa8fa0da8c06e18e0316ec0726b2d34b0/Kernel/Arch/Processor.h#L15-L22

File: Kernel/Arch/aarch64/RegisterState.h [hunk 3]
```diff
@@ -11,35 +11,124 @@
...
-    u64 x[31];     // Saved general purpose registers
+    u64 x[32];     // Saved general purpose registers
     u64 spsr_el1;  // Save Processor Status Register, EL1
     u64 elr_el1;   // Exception Link Register, EL1
     u64 tpidr_el1; // EL0 thread ID
     u64 sp_el0;    // EL0 stack pointer
 
-    FlatPtr userspace_sp() const { return 0; }
+    FlatPtr pc;
+    FlatPtr nzcv;
```
Comment (FireFox317) [inline Kernel/Arch/aarch64/RegisterState.h:27]: This needs a corresponding change in https://github.com/SerenityOS/serenity/blob/master/Kernel/Arch/aarch64/vector_table.S

However, I'm not sure why there are suddenly 32 general purpose registers? I'm pretty sure there are 31 of them.

File: Kernel/Scheduler.cpp [hunk 4]
```diff
@@ -270,11 +282,19 @@ void Scheduler::context_switch(Thread* thread)
         from_thread->set_state(Thread::State::Runnable);
 
 #ifdef LOG_EVERY_CONTEXT_SWITCH
+#    if ARCH(AARCH64)
+    auto const msg = "Scheduler[{}]: {} -> {} [prio={}] {:p} ({:p})";
+
+    dbgln(msg,
+        Processor::current_id(), from_thread->tid().value(),
+        thread->tid().value(), thread->priority(), thread->regs().ip0, thread->regs().ip1);
+#    else
```
Comment (FireFox317) [inline Kernel/Scheduler.cpp:291]: I have a commit locally that just updates the message by getting rid of the code selector, that way we can just both codepaths for x86 and aarch64.

**Comment by FireFox317** [comment]
Also, feel free to hop in the #aarch64 channel on Discord, so we can coordinate things a little better. I have quite some patches layout around that also touch some of this stuff.

File: Userland/Libraries/LibC/sys/arch/aarch64/regs.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2022, Konrad <konradek@github.com>
```
Comment (ADKaster) [inline Userland/Libraries/LibC/sys/arch/aarch64/regs.h:2]: Do you have an actual github.com email address? This looks unusual.

File: Userland/Libraries/LibC/sys/arch/regs.h [hunk 6]
```diff
@@ -12,4 +12,6 @@
 #    include "i386/regs.h"
 #elif ARCH(X86_64)
 #    include "x86_64/regs.h"
+#elif ARCH(AARCH64)
+#    include "aarch64/regs.h" // FIXME
```
Comment (ADKaster) [inline Userland/Libraries/LibC/sys/arch/regs.h:16]: Is this still a FIXME if you implemented it in this commit? One option might be to leave this change to arch/regs.h and the actual implementation of arch/aarch64/regs.h for a second commit, and title the commit named `Everywhere: Single import place for x86 and arm64 registers` as: `LibC+Everywhere: Consistently use sys/arch/regs.h helper file`

File: Userland/Libraries/LibC/sys/arch/aarch64/regs.h [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    u64 sp; // non-general-purpose
+    u64 pc; // non-general-purpose
+
+    // flags
+    u64 nzcv;
+
+    // FIXME: Add FPU registers and Flags
+
+    u64 ip() const { return pc; }
+    void set_ip(u64 value) { pc = value; } // FIXME?
```
Comment (ADKaster) [inline Userland/Libraries/LibC/sys/arch/aarch64/regs.h:73]: A future reader of this code will very likely not understand what this FIXME is about. It should either be removed, or have  some context added. FIXME: Do what?

File: Kernel/Arch/aarch64/Processor.cpp [hunk 8]
```diff
@@ -105,4 +105,9 @@ void Processor::check_invoke_scheduler()
     TODO_AARCH64();
 }
 
+StringView Processor::platform_string()
+{
+    return "aarch64"sv; // FIXME
```
Comment (ADKaster) [inline Kernel/Arch/aarch64/Processor.cpp:110]: What's `FIXME` about this? should the string have more information?

File: Kernel/API/Syscall.h [hunk 9]
```diff
@@ -655,6 +663,8 @@ inline uintptr_t invoke(Function function, T1 arg1, T2 arg2, T3 arg3, T4 arg4)
                  : "r"(x1), "r"(x2), "r"(x3), "r"(x4), "r"(x8)
                  : "memory");
     result = x0;
+#        else
+#            error Unsupported architecture
```
Comment (ADKaster) [inline Kernel/API/Syscall.h:667]: This change to Syscall.h seems independent and should be its own commit

File: Kernel/Arch/aarch64/Processor.h [hunk 10]
```diff
@@ -260,6 +260,8 @@ class Processor {
     FlatPtr init_context(Thread& thread, bool leave_crit);
     static ErrorOr<Vector<FlatPtr, 32>> capture_stack_trace(Thread& thread, size_t max_frames = 0);
 
+    static StringView platform_string();
```
Comment (ADKaster) [inline Kernel/Arch/aarch64/Processor.h:263]: This change to add platform_string() to aarch64/Processor.h seems independent and should be its own commit

File: Kernel/Arch/aarch64/RegisterState.h [hunk 11]
```diff
@@ -11,35 +11,124 @@
...
+        function = x[8];
+        arg1 = x[1];
+        arg2 = x[2];
+        arg3 = x[3];
+        arg4 = x[4];
     }
-    FlatPtr bp() const { TODO_AARCH64(); }
 };
 
 inline void copy_kernel_registers_into_ptrace_registers(PtraceRegisters& ptrace_regs, RegisterState const& kernel_regs)
```
Comment (ADKaster) [inline Kernel/Arch/aarch64/RegisterState.h:55]: The implementation of `copy_kernel_registers_into_ptrace_registers` and the above changes to add implementations of RegisterState's methods are independent and should be in separate commits.

kernel regs --> ptrace regs and ptrace regs --> kernel regs can be in the same commit though

File: Kernel/Arch/aarch64/mcontext.h [hunk 12]
```diff
@@ -13,39 +13,41 @@ extern "C" {
...
+    uint64_t x25;
+    uint64_t x26;
+    uint64_t x27;
+    uint64_t x28;
+    uint64_t x29;
+    uint64_t x30;
+    uint64_t x31;
     uint64_t sp;
     uint64_t pc;
+    uint64_t nzcv;
```
Comment (ADKaster) [inline Kernel/Arch/aarch64/mcontext.h:50]: This rename is its own thing, and can either be grouped with the changes to register state, or handled in a commit that also changes downstream consumers of mcontext in the same commit.

File: Kernel/Arch/x86/RegisterState.h [hunk 13]
```diff
@@ -109,13 +109,21 @@ struct [[gnu::packed]] RegisterState {
...
         arg4 = esi;
-#else
+#elif ARCH(X86_64)
         // The syscall instruction clobbers rcx, so we must use a different calling convention to 32-bit.
         function = rax;
         arg1 = rdx;
         arg2 = rdi;
         arg3 = rbx;
         arg4 = rsi;
+#elif ARCH(AARCH64)
```
Comment (ADKaster) [inline Kernel/Arch/x86/RegisterState.h:119]: This change should not be here. We should never be in Arch/x86/RegisterState.h if ARCH(AARCH64) is defined. this file has `VALIDATE_IS_X86()` at the top.

File: Kernel/CMakeLists.txt [hunk 14]
```diff
@@ -477,6 +477,9 @@ elseif("${SERENITY_ARCH}" STREQUAL "aarch64")
         Arch/aarch64/ScopedCritical.cpp
         Arch/aarch64/SmapDisabler.cpp
         Arch/aarch64/vector_table.S
+
+        Arch/aarch64/ASM_wrapper.h
+        Arch/aarch64/CPU.h
```
Comment (ADKaster) [inline Kernel/CMakeLists.txt:482]: We typically do not have any header files in SOURCES

File: Kernel/Memory/PageDirectory.cpp [hunk 15]
```diff
@@ -52,16 +52,17 @@ ErrorOr<NonnullLockRefPtr<PageDirectory>> PageDirectory::try_create_for_userspac
     }
 #endif
 
+#if ARCH(I386) || ARCH(X86_64)
```
Comment (ADKaster) [inline Kernel/Memory/PageDirectory.cpp:55]: What does this change to PageDirectory achieve? if anything it deserves a comment, but a commit with a descriptive commit message would be better.

File: Kernel/Random.cpp [hunk 16]
```diff
@@ -69,6 +69,8 @@ UNMAP_AFTER_INIT KernelRng::KernelRng()
             current_time += 0x40b2u;
         }
     }
+#elif ARCH(AARCH64)
```
Comment (ADKaster) [inline Kernel/Random.cpp:72]: let's not do this until we have an RNG source for aarch64 to make the extra ifdef branch meaningful.

File: Kernel/Prekernel/init.cpp [hunk 17]
```diff
@@ -17,6 +17,10 @@
 #if ARCH(I386) || ARCH(X86_64)
 #    include <Kernel/Arch/x86/ASM_wrapper.h>
 #    include <Kernel/Arch/x86/CPUID.h>
+#elif ARCH(AARCH64)
+#    include <Kernel/Arch/aarch64/ASM_wrapper.h>
```
Comment (ADKaster) [inline Kernel/Prekernel/init.cpp:21]: Prekernel is not used in Aarch64. why do we need this change?

File: Kernel/Scheduler.cpp [hunk 18]
```diff
@@ -222,10 +222,22 @@ void Scheduler::pick_next()
 
     auto& thread_to_schedule = pull_next_runnable_thread();
     if constexpr (SCHEDULER_DEBUG) {
+#if ARCH(I386) || ARCH(X86_64)
         dbgln("Scheduler[{}]: Switch to {} @ {:#04x}:{:p}",
             Processor::current_id(),
             thread_to_schedule,
             thread_to_schedule.regs().cs, thread_to_schedule.regs().ip());
+#elif ARCH(AARCH64)
+        // TODO
```
Comment (ADKaster) [inline Kernel/Scheduler.cpp:231]: what is todo about this? it looks like it prints more or less the same information now.

File: Kernel/Syscall.cpp [hunk 19]
```diff
@@ -7,8 +7,11 @@
 
 #include <Kernel/API/Syscall.h>
 #include <Kernel/Arch/TrapFrame.h>
-#include <Kernel/Arch/x86/Interrupts.h>
-#include <Kernel/Memory/MemoryManager.h>
+#if ARCH(I386) || ARCH(X86_64)
+#    include <Kernel/Arch/x86/Interrupts.h>
+#elif ARCH(AARCH64)
+#    include <Kernel/Memory/MemoryManager.h>
```
Comment (ADKaster) [inline Kernel/Syscall.cpp:13]: what's going on here? why do we not include MemoryManager on x86 platforms anymore?

File: Kernel/Syscalls/execve.cpp [hunk 20]
```diff
@@ -727,9 +736,14 @@ static Array<ELF::AuxiliaryValue, auxiliary_vector_size> generate_auxiliary_vect
         { ELF::AuxiliaryValue::Gid, (long)gid.value() },
         { ELF::AuxiliaryValue::EGid, (long)egid.value() },
 
+#if ARCH(I386) || ARCH(X86_64)
         { ELF::AuxiliaryValue::Platform, Processor::platform_string() },
         // FIXME: This is platform specific
         { ELF::AuxiliaryValue::HwCap, (long)CPUID(1).edx() },
+#elif ARCH(AARCH64)
+        { ELF::AuxiliaryValue::Platform, Processor::platform_string() },
+        { ELF::AuxiliaryValue::HwCap, 0L }, // TODO
```
Comment (ADKaster) [inline Kernel/Syscalls/execve.cpp:745]: TODO: get HwCap from (???)

File: Kernel/Syscalls/fork.cpp [hunk 21]
```diff
@@ -146,6 +145,45 @@ ErrorOr<FlatPtr> Process::sys$fork(RegisterState& regs)
...
+    child_regs.x22 = regs.x[22];
+    child_regs.x23 = regs.x[23];
+    child_regs.x24 = regs.x[24];
+    child_regs.x25 = regs.x[25];
+    child_regs.x26 = regs.x[26];
+    child_regs.x27 = regs.x[27];
+    child_regs.x28 = regs.x[28];
+    child_regs.fp = regs.x[29];
+    child_regs.lr = regs.x[30];
+    child_regs.rsp = regs.x[31];
```
Comment (ADKaster) [inline Kernel/Syscalls/fork.cpp:180]: rsp is an x86-ism is it not? I thought we had a separate entry in the RegisterState for the stack pointer as `sp_el0`. Why are we grabbing it from the x-register array?

File: Kernel/Syscalls/fork.cpp [hunk 22]
```diff
@@ -146,6 +145,45 @@ ErrorOr<FlatPtr> Process::sys$fork(RegisterState& regs)
...
+
+    child_regs.nzcv = regs.nzcv;
+    child_regs.pc = regs.pc;
+
+//    dbgln_if(FORK_DEBUG, "fork: child will begin executing at {:#04x}:{:p} with stack {:p}, kstack {:p}",
+//        child_regs.cs, child_regs.rip, child_regs.rsp, child_regs.rsp0);
```
Comment (ADKaster) [inline Kernel/Syscalls/fork.cpp:186]: this dbgln should be kept but with proper register names

File: Kernel/Syscalls/mmap.cpp [hunk 23]
```diff
@@ -567,9 +569,11 @@ ErrorOr<FlatPtr> Process::sys$allocate_tls(Userspace<char const*> initial_data,
         auto& tls_descriptor = Processor::current().get_gdt_entry(GDT_SELECTOR_TLS);
         tls_descriptor.set_base(main_thread->thread_specific_data());
         tls_descriptor.set_limit(main_thread->thread_specific_region_size());
-#else
+#elif ARCH(X86_64)
         MSR fs_base_msr(MSR_FS_BASE);
         fs_base_msr.set(main_thread->thread_specific_data().get());
+#elif ARCH(AARCH64)
+        TODO_AARCH64();
```
Comment (ADKaster) [inline Kernel/Syscalls/mmap.cpp:576]: pls annotate this todo to say that we should be setting thread-local data into tpidr_el0

File: Kernel/Thread.h [hunk 24]
```diff
@@ -86,8 +86,57 @@ struct ThreadRegisters {
...
+    FlatPtr x24;
+    FlatPtr x25;
+    FlatPtr x26;
+    FlatPtr x27;
+    FlatPtr x28;
+
+    FlatPtr fp; // x29, frame
+    FlatPtr lr; // x30, link
+
+    FlatPtr rsp; // x31 or zero
```
Comment (ADKaster) [inline Kernel/Thread.h:131]: rsp? this should be sp_el0, no?

File: Kernel/kprintf.cpp [hunk 25]
```diff
@@ -191,3 +196,4 @@ extern "C" void kernelearlyputstr(char const* characters, size_t length)
     for (size_t i = 0; i < length; ++i)
         internal_dbgputch(characters[i]);
 }
+#endif
```
Comment (ADKaster) [inline Kernel/kprintf.cpp:199]: Essentially, all of these syscall file changes should be their own commit. more or less.

File: Userland/Applications/Debugger/main.cpp [hunk 26]
```diff
@@ -11,15 +11,20 @@
 #include <AK/Platform.h>
 #include <AK/StringBuilder.h>
 #include <AK/Try.h>
-#include <LibC/sys/arch/i386/regs.h>
+#include <LibC/sys/arch/regs.h>
```
Comment (ADKaster) [inline Userland/Applications/Debugger/main.cpp:14]: This change should have been in the second commit, that moves everyone off of arch/i386/regs.h

File: Userland/Applications/Debugger/main.cpp [hunk 27]
```diff
@@ -11,15 +11,20 @@
...
 #include <LibLine/Editor.h>
 #include <LibMain/Main.h>
-#include <LibX86/Disassembler.h>
-#include <LibX86/Instruction.h>
+#if ARCH(I386) || ARCH(X86_64)
+#    include <LibX86/Disassembler.h>
+#    include <LibX86/Instruction.h>
+#elif ARCH(AARCH64)
+#    include <LibARM64/Disassembler.h>
+#    include <LibARM64/Instruction.h>
```
Comment (ADKaster) [inline Userland/Applications/Debugger/main.cpp:26]: creating LibARM64 and using it in disasm should be distinct commits

File: Userland/Libraries/LibARM64/Instruction.cpp [hunk 28]
```diff
@@ -0,0 +1,2539 @@
...
+    d.lock_prefix_allowed = lock_prefix_allowed;
+
+    if ((format > __BeginFormatsWithRMByte && format < __EndFormatsWithRMByte) || format == MultibyteWithSlash)
+        d.has_rm = true;
+    else
+        d.opcode_has_register_index = opcode_has_register_index(op);
+
+    switch (format) {
+    case OP_RM8_imm8:
+    case OP_RM16_imm8:
```
Comment (ADKaster) [inline Userland/Libraries/LibARM64/Instruction.cpp:52]: This file is clearly unrelated to aarch64. Let's not just copy paste Libx86 files and call them "LibARM64". SSE is not a thing in aarch64. these instruction formats are not valid for aarch64.

**Review by ADKaster — CHANGES_REQUESTED** [review]
I left some comprehensive comments, but the main takeaway is "split these commits up". 

LibARM64 seems to be a copy pasta of LibX86, which we absolutely should not do. If there's no actual arm64 code in there, let's wait on that library until there's some actual arm64 instruction decoding going on.

The changes to kernel files are largely independent. the changes to utilities can be done piecemeal. Some of the changes I'm not convinced are required in order to compile, and the register naming is a bit suspect, as was already pointed out.

File: Kernel/Arch/aarch64/RegisterState.h (hunk 3 above)
Comment (konradekk) [reply Kernel/Arch/aarch64/RegisterState.h:27]: checked in the docs (cf. [AArch64 special registers](https://developer.arm.com/documentation/den0024/a/ARMv8-Registers/AArch64-special-registers)), youʼre right here! 🙇🏻‍♂️

i was thinking of this facultative use of sp register but it seems itʼs only a part of encoding scheme, so this approach with ‘x31ʼ might be more of greater use in instruction decoder for assembly; iʼll remove it from here.

File: Userland/Libraries/LibC/sys/arch/aarch64/regs.h (hunk 5 above)
Comment (konradekk) [reply Userland/Libraries/LibC/sys/arch/aarch64/regs.h:2]: hahah! yup! is it okay to use <konrad@serenityos.org> without having mail (alias) there? 😝

File: Userland/Libraries/LibC/sys/arch/aarch64/regs.h (hunk 7 above)
Comment (konradekk) [reply Userland/Libraries/LibC/sys/arch/aarch64/regs.h:73]: definitely nothing to be done here—just a leftover! 🫢

File: Userland/Libraries/LibC/sys/arch/regs.h (hunk 6 above)
Comment (konradekk) [reply Userland/Libraries/LibC/sys/arch/regs.h:16]: awww! 🫢 iʼll take the approach for the change too!

File: Kernel/Scheduler.cpp (hunk 4 above)
Comment (konradekk) [reply Kernel/Scheduler.cpp:291]: iʼll guard that with non aarch64 path then (leaving the change for you).

File: Userland/Libraries/CMakeLists.txt (hunk 1 above)
Comment (konradekk) [reply Userland/Libraries/CMakeLists.txt:66]: will do!

File: Kernel/Arch/aarch64/Processor.cpp (hunk 8 above)
Comment (konradekk) [reply Kernel/Arch/aarch64/Processor.cpp:110]: iʼll provide more details (it should be just extracted from registers). 🙇🏻‍♂️

File: Kernel/Arch/aarch64/RegisterState.h (hunk 11 above)
Comment (konradekk) [reply Kernel/Arch/aarch64/RegisterState.h:55]: i understand that each commit should yield in coherent (i.e. runnable) state, right?

File: Kernel/Arch/x86/RegisterState.h (hunk 13 above)
Comment (konradekk) [reply Kernel/Arch/x86/RegisterState.h:119]: hah! very true! 🫣

File: Kernel/CMakeLists.txt (hunk 14 above)
Comment (konradekk) [reply Kernel/CMakeLists.txt:482]: oh!

File: Kernel/Prekernel/init.cpp (hunk 17 above)
Comment (konradekk) [reply Kernel/Prekernel/init.cpp:21]: shall (eventually) go under `x86`?

File: Kernel/Random.cpp (hunk 16 above)
Comment (konradekk) [reply Kernel/Random.cpp:72]: i was working on it but it seems itʼs not really a thing for raspberry processor… 😒
(it has some additional hardware but i donʼt know yet how to access it)

iʼll remove that then. 🙇🏻‍♂️

File: Kernel/Syscall.cpp (hunk 19 above)
Comment (konradekk) [reply Kernel/Syscall.cpp:13]: 🫢

File: Kernel/Syscalls/execve.cpp (hunk 20 above)
Comment (konradekk) [reply Kernel/Syscalls/execve.cpp:745]: processor? thereʼre cpuid counterparts… but that might be to specific (i still planned to use that information). got some ideas on what should i do here exactly?

File: Kernel/Syscalls/fork.cpp (hunk 21 above)
Comment (konradekk) [reply Kernel/Syscalls/fork.cpp:180]: related to firefox37 comments; to be removed.

File: Userland/DevTools/HackStudio/CMakeLists.txt [hunk 29]
```diff
@@ -55,5 +55,5 @@ set(GENERATED_SOURCES
 )
 
 serenity_app(HackStudio ICON app-hack-studio)
-target_link_libraries(HackStudio PRIVATE LibWebView LibWeb LibMarkdown LibGUI LibCpp LibGfx LibCore LibVT LibDebug LibX86 LibDiff LibShell LibSymbolication LibSyntax LibRegex LibSQL LibConfig LibCore LibCoredump LibDesktop LibIPC LibJS LibMain LibThreading)
+target_link_libraries(HackStudio PRIVATE LibWebView LibWeb LibMarkdown LibGUI LibCpp LibGfx LibCore LibVT LibDebug LibX86 LibARM64 LibARM64 LibDiff LibShell LibSymbolication LibSyntax LibRegex LibSQL LibConfig LibCore LibCoredump LibDesktop LibIPC LibJS LibMain LibThreading)
```
Comment (konradekk) [inline Userland/DevTools/HackStudio/CMakeLists.txt:58]: oh!
